doc/api/single-executable-applications.md

```diff
@@ -177,14 +181,33 @@ The configuration currently reads the following top-level fields:
 {
   "main": "/path/to/bundled/script.js",
   "output": "/path/to/write/the/generated/blob.blob",
-  "disableExperimentalSEAWarning": true // Default: false
+  "disableExperimentalSEAWarning": true, // Default: false
+  "useSnapshot": true  // Default: false
 }
 ```
 
 If the paths are not absolute, Node.js will use the path relative to the
 current working directory. The version of the Node.js binary used to produce
 the blob must be the same as the one to which the blob will be injected.
 
+### Startup snapshot support
+
+When `useSnapshot` is set to true in the configuration, during the generation
+of the single executable preparation blob, Node.js will run the `main` script
+to generate a startup snapshot. The script must invoke
+[`v8.startupSnapshot.setDeserializeMainFunction()`][] to set up the entry point.
+The generated startup snapshot would be part of the preparation blob and get
+injected into the final executable. When the single executable application is
```

Comment:
Would it be possible to generate and use the snapshot in a way that doesn't require users to use the `v8.startupSnapshot.setDeserializeMainFunction()` function in their code? I think that would help in making this easier to use because that way users wouldn't be required to change the SEA code to use this feature, they just need to set `useSnapshot` to `true` and that should be it AFAICT.

Reply:
Using snapshot should involve intentional effort to split the application code into initialization code that can be run during snapshot-building time and application code that can be run during run-time. When `useSnapshot` is true, the script is run *at build time* to produce the snapshot, and it's expected to run to completion. So if user does not configure additional code to be run at run-time...then we don't have anything to run when the SEA is launched.

If what we want is not having to do code changes, one way to do it is to add another `snapshot_script` field so that a separate snapshot script is run at build time and the main script is left to be run at run-time.  But I don't think that's necessarily better than requiring `v8.startupSnapshot.setDeserializeMainFunction()`, because then it's harder for the main script to retrieve states set up by the snapshot script without, um, using any JS API explicitly (unless the two scripts use some global objects to communicate, which is not ideal). And when snapshot is used it's usually useful or necessary to be aware about it and fix up the states, or do things differently, it's not designed to be something that can just be configured with a boolean and be it IMO.

Reply:
ping @RaisinTen does that answer your question?